crates/provider/src/native.rs

```diff
@@ -296,19 +289,83 @@ impl<T: FileSystem + Send + Sync> Provider for NativeProvider<T> {
         Ok(format!("tail -f {}/{}.log", self.tmp_dir, name))
     }
 
+    // TODO: Add test
     async fn pause(&self, node_name: &str) -> Result<(), ProviderError> {
-        // TODO: impl using run_command
+        let process_id: Result<u32, ProviderError> = match self.process_map.get(node_name) {
+            Some(process) => Ok(process.pid),
+            None => Err(ProviderError::MissingNodeInfo(
+                node_name.to_owned(),
+                "process".into(),
+            )),
+        };
+
+        let _ = self
+            .run_command(
+                [format!("kill -TSTP {}", process_id.unwrap())].to_vec(),
+                NativeRunCommandOptions {
+                    is_failure_allowed: true,
+                },
+            )
+            .await?;
         Ok(())
     }
 
+    // TODO: Add test
     async fn resume(&self, node_name: &str) -> Result<(), ProviderError> {
-        // TODO: impl using run_command
+        let process_id: Result<u32, ProviderError> = match self.process_map.get(node_name) {
+            Some(process) => Ok(process.pid),
+            None => Err(ProviderError::MissingNodeInfo(
+                node_name.to_owned(),
+                "process".into(),
+            )),
+        };
+
+        let _ = self
+            .run_command(
+                [format!("kill -CONT {}", process_id.unwrap())].to_vec(),
+                NativeRunCommandOptions {
+                    is_failure_allowed: true,
+                },
+            )
+            .await?;
         Ok(())
     }
 
-    async fn restart(&mut self, node_name: &str, after_secs: u16) -> Result<bool, ProviderError> {
-        // TODO: impl using run_command
-        // use &mut here since we mostly need to update the the pid/maps
+    // TODO: Add test
+    async fn restart(&mut self, node_name: &str, _after_secs: u16) -> Result<bool, ProviderError> {
+        let process: Result<Process, ProviderError> = match self.process_map.get(node_name) {
+            Some(pr) => Ok(pr.clone()),
+            None => Err(ProviderError::MissingNodeInfo(
+                node_name.to_owned(),
+                "process".into(),
+            )),
+        };
+
+        let mut process = process?;
+
+        let pid = process.pid;
+        let command = process.command.clone();
+
+        let _ = self
+            .run_command(
+                [format!("kill -9 {}", pid)].to_vec(),
+                NativeRunCommandOptions {
+                    is_failure_allowed: true,
+                },
+            )
+            .await?;
+
+        let mapped_env: HashMap<String, String> = env::vars().map(|(k, v)| (k, v)).collect();
+
+        let res = Command::new(command).arg("-c").envs(&mapped_env).spawn();
+
+        let restarted_process = match res {
+            Ok(process) => process,
+            Err(e) => return Err(ProviderError::ErrorSpawningNode(e.to_string())),
+        };
+
+        process.pid = restarted_process.id().unwrap();
+
```

Comment:
This will not work  as expected, since the `clone` will create a new `Process` and you are then shadowing as mut but never changing the underline process in the map. Look this [example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f08d3da94aafa1d164c5aa4521dd0f6b)

You need to use `[get_mut](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.get_mut)` in order to get a mutable ref. Ping me and we can touch base on this one.